### PR TITLE
passthrough readOnly to subpath

### LINKED ISSUE #62752: CreateContainerConfigError : failed to prepare subPath for volumeMount (possibly only for readOnly volumes)
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

I get a "failed to prepare subPath for volumeMount" error (CreateContainerConfigError) when trying to deploy on a 1.9.6-gke.1 node.

**What you expected to happen**:

I should be able to use subPaths on volume mounts, especially because the entire disk is read-only.

**How to reproduce it (as minimally and precisely as possible)**:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
    name: fomopop-website-deployment
spec:
    replicas: 3
    template:
        spec:
            containers:
                - name: fomopop-website
                  image: gcr.io/fomopop-disco-beach/fomopop-website:2018-04-17-12-05
                  volumeMounts:
                      - name: fomopop-sheet-music
                        subPath: website/content
                        mountPath: /app/website/content
                      - name: fomopop-sheet-music
                        subPath: website/thumbs
                        mountPath: /app/website/thumbs
                      - name: fomopop-sheet-music
                        subPath: website/site/accounts
                        mountPath: /app/website/accounts
            volumes:
                - name: fomopop-sheet-music
                  gcePersistentDisk:
                      pdName: fomopop-sheet-music-2018-04-17-12-08
                      fsType: ext4
                      readOnly: true
```

**Anything else we need to know?**:
Everything works fine on 1.9.3-gke.0 nodes.

Based on one error log message, it seems like there is a regression in previous defined/undefined behavior with regards to readOnly volumes. The **volume** I am attempting to mount subPaths from is readOnly. However, I previously _did not_ specify that the individual **volumeMounts** should be readOnly; I assumed it was obvious/clear that if the source volume is readOnly, then the mount should be readOnly.

I've tried to add `readOnly: true` to the volumeMounts, but that gives no improvement. The logs still indicate that it's trying to mount it as read-write when it (probably?) shouldn't.

Error logs from node:
```
2018-04-17 14:16:55.000 PDT
E0417 21:16:55.296480 1331 kubelet_pods.go:248] failed to prepare subPath for volumeMount "fomopop-sheet-music" of container "fomopop-website": error mounting /var/lib/kubelet/pods/4d1cf26c-4283-11e8-8e1a-42010a800150/volumes/kubernetes.io~gce-pd/fomopop-sheet-music/website/content: mount failed: exit status 32

2018-04-17 14:17:33.000 PDT
E0417 21:17:33.301247 1331 pod_workers.go:186] Error syncing pod 4d1cf26c-4283-11e8-8e1a-42010a800150 ("fomopop-website-deployment-74cbf44bb9-8vgxl_default(4d1cf26c-4283-11e8-8e1a-42010a800150)"), skipping: failed to "StartContainer" for "fomopop-website" with CreateContainerConfigError: "failed to prepare subPath for volumeMount \"fomopop-sheet-music\" of container \"fomopop-website\""

2018-04-17 14:17:46.000 PDT
E0417 21:17:46.306213 1331 kuberuntime_manager.go:734] container start failed: CreateContainerConfigError: failed to prepare subPath for volumeMount "fomopop-sheet-music" of container "fomopop-website"

2018-04-17 14:18:13.000 PDT
E0417 21:18:13.296460 1331 kuberuntime_manager.go:734] container start failed: CreateContainerConfigError: failed to prepare subPath for volumeMount "fomopop-sheet-music" of container "fomopop-website"

2018-04-17 14:18:37.000 PDT
E0417 21:18:37.295800 1331 pod_workers.go:186] Error syncing pod 4d1cf26c-4283-11e8-8e1a-42010a800150 ("fomopop-website-deployment-74cbf44bb9-8vgxl_default(4d1cf26c-4283-11e8-8e1a-42010a800150)"), skipping: failed to "StartContainer" for "fomopop-website" with CreateContainerConfigError: "failed to prepare subPath for volumeMount \"fomopop-sheet-music\" of container \"fomopop-website\""

2018-04-17 14:21:42.000 PDT
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/4d1cf26c-4283-11e8-8e1a-42010a800150/volume-subpaths/fomopop-sheet-music/fomopop-website/0 --scope -- mount -o remount /proc/1331/fd/6 /var/lib/kubelet/pods/4d1cf26c-4283-11e8-8e1a-42010a800150/volume-subpaths/fomopop-sheet-music/fomopop-website/0

2018-04-17 14:23:07.000 PDT
mount: cannot remount /var/lib/kubelet/pods/4d1cf26c-4283-11e8-8e1a-42010a800150/volumes/kubernetes.io~gce-pd/fomopop-sheet-music/website/content read-write, is write-protected

2018-04-17 14:24:40.000 PDT
E0417 21:24:40.299188 1331 kuberuntime_manager.go:734] container start failed: CreateContainerConfigError: failed to prepare subPath for volumeMount "fomopop-sheet-music" of container "fomopop-website"

2018-04-17 14:25:24.000 PDT
E0417 21:25:24.268653 1331 mount_linux.go:145] Mount failed: exit status 32
```
**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.6", GitCommit:"6260bb08c46c31eea6cb538b34a9ceb3e406689c", GitTreeState:"clean", BuildDate:"2017-12-21T06:34:11Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.6-gke.1", GitCommit:"cb151369f60073317da686a6ce7de36abe2bda8d", GitTreeState:"clean", BuildDate:"2018-04-07T22:06:59Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
Google Kubernetes Engine (GKE)

- OS (e.g. from /etc/os-release):
```
BUILD_ID=10323.69.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=94bdaa2af49562e7aba0168cae050911d105f5cb
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=65
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=65
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```

- Kernel (e.g. `uname -a`):
```
Linux gke-fomopop-website-clu-small-preempt-31ce5ba7-jrm6 4.4.111+ #1 SMP Thu Apr 5 21:21:21 PDT 2018 x86_64 Intel(
R) Xeon(R) CPU @ 2.60GHz GenuineIntel GNU/Linux
```

- Install tools:
- Others:

@kubernetes/sig-storage-bugs
/sig storage

### PR DESCRIPTION
**Pull request by msau42**
**What this PR does / why we need it**:
If a volume is mounted as readonly, or subpath volumeMount is configured as readonly, then the subpath bind mount should be readonly.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62752

**Special notes for your reviewer**:

**Release note**:

```release-note
Fixes issue where subpath readOnly mounts failed
```

### CODE REVIEW

**Comment by msau42** [comment]
I am still working on e2e tests

**Comment by msau42** [comment]
cc @andyzhangx 
probably something similar needs to be done for windows

**Comment by msau42** [comment]
cc @kubernetes/sig-storage-pr-reviews

**Comment by msau42** [comment]
Hm this is still not working:

```
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: I0424 01:12:55.957265    1327 mount_linux.go:143] Mounting cmd (systemd-run) with arguments ([--description=Kubernetes transient mount for /var/lib/
kubelet/pods/f15795a3-475b-11e8-b4d0-42010a800002/volume-subpaths/test-volume/test-container-subpath-emptydir-k7t9/0 --scope -- mount -o remount,ro /proc/1327/fd/34 /var/lib/kubelet/pods/f15795a3-475b-11e8-b4d0-
42010a800002/volume-subpaths/test-volume/test-container-subpath-emptydir-k7t9/0])
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: E0424 01:12:55.964589    1327 mount_linux.go:148] Mount failed: exit status 32
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: Mounting command: systemd-run
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/f15795a3-475b-11e8-b4d0-42010a800002/volume-subpaths/test-vol
ume/test-container-subpath-emptydir-k7t9/0 --scope -- mount -o remount,ro /proc/1327/fd/34 /var/lib/kubelet/pods/f15795a3-475b-11e8-b4d0-42010a800002/volume-subpaths/test-volume/test-container-subpath-emptydir-k
7t9/0
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: Output: Running scope as unit: run-rcf0ee9fe5efe41e9a1bb0a8a752b6333.scope
Apr 24 01:12:55 e2e-test-msau-minion-group-nchn kubelet[1327]: mount: /var/lib/kubelet/pods/f15795a3-475b-11e8-b4d0-42010a800002/volume-subpaths/test-volume/test-container-subpath-emptydir-k7t9/0 is busy
```

**Comment by msau42** [comment]
pushed some e2e tests, but they're all failing

looks like bind mounting read only is not working. will debug more tomorrow.

**Comment by andyzhangx** [comment]
@msau42 Actually `readonly` flag is not respected for windows dir mount, still in investigation, anyway I filed a issue: https://github.com/kubernetes/kubernetes/issues/63051

File: pkg/util/mount/mount.go [hunk 1]
```diff
@@ -274,7 +276,7 @@ func IsNotMountPoint(mounter Interface, file string) (bool, error) {
 // The list equals:
 //   options - 'bind' + 'remount' (no duplicate)
 func isBind(options []string) (bool, []string) {
-	bindRemountOpts := []string{"remount"}
+	bindRemountOpts := []string{"bind", "remount"}
```
Comment (msau42) [inline pkg/util/mount/mount.go]: This is what I had to do to make it work for the case that the volume is writable, but the subpath is read only.  

The original behavior of bind mount is: If bind mount is read only, and the underlying volume is writable, then this remount will also remount the original volume as read only. For this fix, if you also pass in the "bind" option to the remount, then it will make only the bind mount read only, and will not change the original volume mount.

We need to be sure this behavior change is ok for normal bind mounts that volume plugins do.

File: pkg/util/mount/mount_linux.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ const (
 	// syscall.Openat flags used to traverse directories not following symlinks
 	nofollowFlags = unix.O_RDONLY | unix.O_NOFOLLOW
 	// flags for getting file descriptor without following the symlink
-	openFDFlags = unix.O_NOFOLLOW | unix.O_PATH
+	openFDFlags = unix.O_RDONLY | unix.O_NOFOLLOW | unix.O_PATH
```
Comment (msau42) [inline pkg/util/mount/mount_linux.go]: This may not be needed. Haven't tested without it yet

File: test/e2e/storage/subpath.go [hunk 3]
```diff
@@ -271,6 +271,18 @@ var _ = utils.SIGDescribe("Subpath", func() {
 				}
 				testSubpathReconstruction(f, pod, true)
 			})
+
+			It("should support readOnly specified in the volumeMount", func() {
```
Comment (msau42) [inline test/e2e/storage/subpath.go]: Also need test cases for when the volume source is read only.  For inline volumes and PV volumes.

**Comment by msau42** [comment]
I got it working, see comments above.

File: pkg/util/mount/mount_linux.go (hunk 2 above)
Comment (msau42) [reply pkg/util/mount/mount_linux.go]: yup removed

File: test/e2e/storage/subpath.go (hunk 3 above)
Comment (msau42) [reply test/e2e/storage/subpath.go]: added

**Comment by msau42** [comment]
containerized kubelet still needs to be tested
/hold

**Comment by msau42** [comment]
Opened #63156 for gce pd test failure.

/retest

File: pkg/util/mount/exec_mount_test.go [hunk 4]
```diff
@@ -66,7 +66,7 @@ func TestBindMount(t *testing.T) {
 			expectedArgs = []string{"-t", fsType, "-o", "bind", sourcePath, destinationPath}
 		case 2:
 			// mount -t fstype -o "remount,opts" source target
-			expectedArgs = []string{"-t", fsType, "-o", "remount," + strings.Join(mountOptions, ","), sourcePath, destinationPath}
+			expectedArgs = []string{"-t", fsType, "-o", "bind,remount," + strings.Join(mountOptions, ","), sourcePath, destinationPath}
```
Comment (davidz627) [inline pkg/util/mount/exec_mount_test.go]: is this related to the readOnly fix?

File: pkg/util/mount/mount.go (hunk 1 above)
Comment (davidz627) [reply pkg/util/mount/mount.go]: just to confirm, previously we were still bind-mounting right? just not explicitly passing in this option, which as far as we know only changes behavior in this specific case?

File: pkg/util/mount/mount_linux_test.go [hunk 5]
```diff
@@ -1220,6 +1221,29 @@ func TestBindSubPath(t *testing.T) {
...
+				subpath := filepath.Join(volpath, "dir0")
+				return nil, volpath, subpath, os.MkdirAll(subpath, defaultPerm)
+			},
+			expectError: false,
+			readOnly:    true,
+		},
+		{
+			name: "subpath-file-readonly",
+			prepare: func(base string) ([]string, string, string, error) {
+				volpath, _ := getTestPaths(base)
```
Comment (davidz627) [inline pkg/util/mount/mount_linux_test.go]: In these test cases the volpath is not readOnly right? Should we add two more tests here that test ReadOnly Subpath Mount for dir/file for when volPath is readOnly?

File: pkg/util/mount/mount_linux_test.go [hunk 6]
```diff
@@ -1269,12 +1294,39 @@ func TestBindSubPath(t *testing.T) {
...
+		if mp.Path == bindPathTarget {
+			foundReadOnly := false
+			for _, opts := range mp.Opts {
+				if opts == "ro" {
+					foundReadOnly = true
+					break
+				}
+			}
+			if expectedReadOnly != foundReadOnly {
+				return fmt.Errorf("expected readOnly %v for mount point %v", expectedReadOnly, bindPathTarget)
```
Comment (davidz627) [inline pkg/util/mount/mount_linux_test.go]: expected... "got..."? Doesn't seem strictly necessary in this case but it might make it slightly clearer

File: test/e2e/storage/subpath.go [hunk 7]
```diff
@@ -271,6 +272,54 @@ var _ = utils.SIGDescribe("Subpath", func() {
...
+
+				// Write the file in the volume from container 1
+				setWriteCommand(filePathInVolume, &pod.Spec.Containers[1])
+
+				By("Running a pod to initialize the volume data")
+				testReadFile(f, filePathInSubpath, pod, 0)
+
+				// Set volume source to read only
+				pod.Spec.Volumes[0].VolumeSource = *roVol
+				clearSubpathPodCommands(pod)
```
Comment (davidz627) [inline test/e2e/storage/subpath.go]: not super clear on why we need to do this

File: test/e2e/storage/subpath.go [hunk 8]
```diff
@@ -271,6 +272,54 @@ var _ = utils.SIGDescribe("Subpath", func() {
...
+
+				// Initialize content in the volume
+				// Create the directory
+				setInitCommand(pod, fmt.Sprintf("mkdir -p %s", subPathDir))
+
+				// Write the file in the volume from container 1
+				setWriteCommand(filePathInVolume, &pod.Spec.Containers[1])
+
+				By("Running a pod to initialize the volume data")
+				testReadFile(f, filePathInSubpath, pod, 0)
```
Comment (davidz627) [inline test/e2e/storage/subpath.go]: why are we testReadFile-ing here?

File: test/e2e/storage/subpath.go [hunk 9]
```diff
@@ -718,6 +785,15 @@ func (s *gcepdSource) createVolume(f *framework.Framework) volInfo {
 	}
 }
 
+func (s *gcepdSource) readOnlyVolume() *v1.VolumeSource {
+	return &v1.VolumeSource{
+		PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
```
Comment (davidz627) [inline test/e2e/storage/subpath.go]: should this be GCEPDSource? Or maybe this should have a gcepdPVCSource type receiver. Or maybe the naming is just inconsistent

File: test/e2e/storage/subpath.go [hunk 10]
```diff
@@ -718,6 +785,15 @@ func (s *gcepdSource) createVolume(f *framework.Framework) volInfo {
 	}
 }
 
+func (s *gcepdSource) readOnlyVolume() *v1.VolumeSource {
```
Comment (davidz627) [inline test/e2e/storage/subpath.go]: these functions are "creating" a readOnlyVolume right? The naming is a little ambiguous, at first I thought it was an "isReadOnlyVolume" type function but it doesn't seem like that.

Comment (davidz627) [reply test/e2e/storage/subpath.go]: createReadOnlyVolumeSource?

File: test/e2e/storage/subpath.go (hunk 9 above)
Comment (davidz627) [reply test/e2e/storage/subpath.go]: ack

File: test/e2e/storage/subpath.go (hunk 8 above)
Comment (davidz627) [reply test/e2e/storage/subpath.go]: yea it's not super clear here, if the functions are being reused for different purposes maybe the right move here is to rename the function?

File: pkg/util/mount/mount_linux_test.go (hunk 5 above)
Comment (davidz627) [reply pkg/util/mount/mount_linux_test.go]: but could it have different behavior whether the permissions on the volume are read only or not? Almost like doing the same test but in a "different environment"

File: test/e2e/storage/subpath.go (hunk 8 above)
Comment (msau42) [reply test/e2e/storage/subpath.go]: updated in new commit

File: pkg/util/mount/mount.go (hunk 1 above)
Comment (davidz627) [reply pkg/util/mount/mount.go]: So this change actually "sort of" deviates from the article you linked, because it looks like in theory we should be able to get a read-only bindmount by doing (according to this article linked):
```
mount -o bind /vital_data /untrusted_container/vital_data
mount -o remount,ro /untrusted_container/vital_data
```

which we had before. However, that actually has the behavior of modifying the original mount point to also be read only, so what you've done is changed the command to the following to fix that:

```
mount -o bind /vital_data /untrusted_container/vital_data
mount -o bind,remount,ro /untrusted_container/vital_data
```

Let me know if that's a flawed understanding.

Comment (davidz627) [reply pkg/util/mount/mount.go]: Interesting thing I found:
https://unix.stackexchange.com/questions/198590/what-is-a-bind-mount

With Linux, the simple way:

mount --bind /some/where /mnt/readonly
mount -o remount,ro,bind /mnt/readonly
This leaves a short interval of time during which /mnt/readonly is read-write. If this is a security concern, first create the bind mount in a directory that only root can access, make it read-only, then move it to a public mount point. In the snippet below, note that it's important that /root/private (the directory above the mount point) is private; the original permissions on /root/private/mnt are irrelevant since they are hidden behind the mount point.

mkdir -p /root/private/mnt
chmod 700 /root/private
mount --bind /some/where /root/private/mnt
mount -o remount,ro,bind /root/private/mnt
mount --move /root/private/mnt /mnt/readonly

Comment (msau42) [reply pkg/util/mount/mount.go]: Ah ah yes you're right.  I totally forgot to explain the issue I hit when it wasn't specifying the "bind" option in the remount.

Because subpath is a subdirectory of the volume, when you do the remount on the base volume as readonly, that remount will fail because of a "device is busy" error.  And I think that is because the subpath is a nested mount inside the base volume.

File: pkg/util/mount/mount_linux_test.go (hunk 5 above)
Comment (msau42) [reply pkg/util/mount/mount_linux_test.go]: added

File: pkg/util/mount/mount.go (hunk 1 above)
Comment (davidz627) [reply pkg/util/mount/mount.go]: Interesting, that's not the problem I expected, might be worth putting a comment above this line with a quick explanation or link to a description of why its necessary. Could be easily confused for a bug or misunderstood in the future.

**Comment by davidz627** [comment]
One last comment about adding a comment, besides that
/lgtm
needs squash

**Comment by msau42** [comment]
This is ready for review. @jsafrane ptal

**Comment by jsafrane** [comment]
/approve 
Thanks!
